Lesson

kadikraman opened the pull request:
- [x] - introduction
- [x] - componentWillMount
- [x] - componentDidMount
- [x] - componentWillReceiveProps
- [x] - shouldComponentUpdate
- [x] - componentWillUpdate
- [x] - componentDidUpdate
- [x] - componentWillUnmount
- [x] - summary (possibly a summarising table)

Commit:
Add initial module layout and shouldComponentUpdate section

Commit:
Add componentWillMount, componentDidMount and componentWillUnmount

Commit:
Add componentWillReceiveProps, componentWillUpdate and componentDidUpdate

Commit:
Add introduction and summary

Commit:
Use javascript syntax highlighting in code snippets

Commit:
Update license and contibuting files to the latest version

thomastuts commented:
Hi @kadikraman, thanks for the PR!

While this lesson provides a handy overview of the lifecycle methods, I feel like it currently reads more like a reference rather than a lesson we guide students through. We try to craft a little story/journey that we take students through together, including a supporting narrative that helps frame the lesson and its objectives. That's kind of lacking here, and it results in the content feeling 'dry' — while that's certainly useful for some cases (i.e. experienced devs who need the reference and not the 'fluff'), it's not completely what we're aiming for with these lessons.

We should try and make the content a little more personal and friendly — something students can relate to. We do so through the use of metaphors, narratives, humor, as well as the 'we're in this together' feeling we create using second person ('you') and first person plural ('we') a lot.

I like the idea of recapping everything at the end — we should definitely keep that in (but maybe in a more condensed format? TBD).

Let me know if you have any questions, and thanks again!

Commit:
Rewrite module in less technical terms

kadikraman commented:
@thomastuts @AnnJohn I've rewritten the module in less technical terms and it's more of an high-level overview rather than a detailed one now. I've removed all the code examples as well - they can be used in the deep dive in the next few modules instead.
Let me know what you think!

thomastuts commented on README.md:
```diff
@@ -7,11 +7,60 @@
 
-This lesson is a very high-level over view of the React component lifecycle. We
-want to walk students through the whole shebang.
+As you've learned previously, a React component has two sets of properties: **props** and **state**. Props are given to the component by it's parent. You can think of props as an external influence that the component has no control over - like the weather. The other set of properties the component has is called the state. A component's state is internal to the component and it can change when the ...
```
> As you've learned previously

We try to decouple lessons as much as possible, so in general, it's best to not refer to any 'previous' lessons — it makes sense in this course, but lessons might be pulled out and reordered.

> and it can change when the props changing
> ... For example, if the weather went from sunny to raining, you would put on a raincoat and wellies.
> In React component terms, your state could go from "wearing a jacket" to "wearing a raincoat and wellies".

State changes on prop changes should never really happen, we use computed values for those instead of adding it to the state — can we remove this? Might be a little confusing.

TIL what wellies are! 👢

kadikraman replied:
But in that example, it is a computed value, I think? From your [reference reading](http://busypeoples.github.io/post/react-component-lifecycle/): 

> **componentWillReceiveProps** enables to update the state depending on the existing and upcoming props, without triggering another rendering

The official [Facebook article](https://facebook.github.io/react/docs/component-specs.html#updating-componentwillreceiveprops) has an example where state is updated in response to props changing.

In the bit above, I meant something like:

```
componentWillReceiveProps: function(nextProps) {
  this.setState({
    clothes: nextProps.weather === 'raining' ? 'jacket' : 'raincoat'
  });
}
```

So it is really a "computed value". But I didn't want to put the actual code in until next modules as not to be too technical from the onset.

Do you still think I should remove it?

thomastuts replied:
The Facebook code (below for reference) is a good example of setting state using props because it's a computed value based on the _difference_ of the old and new props.

``` js
componentWillReceiveProps: function(nextProps) {
  this.setState({
    likesIncreasing: nextProps.likeCount > this.props.likeCount
  });
}
```

In your example, you're basically only checking the new props - this stuff could be done directly in the `render()` method:

``` js
render() {
  const whatToWear = this.props.weather === 'raining' ? 'jacket' : 'raincoat';

  return (
    <p>I should be wearing a {whatToWear} today.</p>
  );
}
```

Doing something like this in the Facebook example would be impossible, because at that point you only have access to the current props - necessitating the need for using state. If we translate the Facebook example to our current weather narrative, we could have an example where we check if the temperature has risen or dropped (or maybe even with how many degrees) and set the state accordingly.

I'm not sure if I'm making a whole lot of sense here - let me know if you still have any questions!

kadikraman replied:
Ah yeah that makes sense. I guess I would never really use React in such a way, I got a bit keen with the "explain it in super simple friendly terms", I'm usually not too shabby at analogies, but it's a pretty technical topic ☀️ - You're right though, I'll get rid of it.

thomastuts commented on README.md:
```diff
@@ -7,11 +7,60 @@
 
-It might be helpful to break the cycle down into mounting (happens once),
-rendering (happens a bazillion times), and unmounting (happens once) phases.
+React components go through what we call a **component lifecycle**. This is broadly divided into three parts: **creation**, **updating** and **deletion**. The React framework was actually created by the social networking site Facebook because they needed a more *reactive* (see what they did there?) way to display information to their users. This means that every single thing you see on Facebook i ...
```
> The React framework was actually created by the social networking site Facebook because they needed a more _reactive_ (see what they did there?) way to display information to their users.

I think we can be a bit less verbose here, e.g. _React was created to create reactive UI's_ or something? Maybe it's also worth decoupling this from the specific example of Facebook, but keep the ChatBox analogy? I really like how you use that example to quickly illustrate the lifecycle. 👍

thomastuts commented on README.md:
```diff
@@ -7,11 +7,60 @@
+When the component is initially created, it gets "mounted" onto the DOM. It sounds more complicated than it is: essentially the component figures out its initial state and renders its initial HTML onto the page. At the mounting stage, there are two *lifecycle hooks* you can use: **componentWillMount** and **componentDidMount**. Remember, the only required method for a React component is the `rend ...
+
+## Unmounting
+At the unmounting stage, the component gets deleted and cleared out of the page. The only lifecycle hook at this stage is **componentWillUnmount**, which is called just before the component gets deleted. This is used to clear out any stuff set up in **componentDidMount**. For example, if you had a component that displays the weather data in your home town, you might have set it up to re-fetch the ...
```
> We will look at a detailed example of how it's done in the next module.

See lesson decoupling note above, we generally try to avoid this

thomastuts commented on README.md:
```diff
@@ -7,11 +7,60 @@
+**componentDidUpdate** is called just after the new component had been rendered. You will have access to the previous props and state as well as the current ones, and you can use this method to update any third party libraries if they happen to need an update due to the re-render.
+
+## Summary
+If you need a lifecycle hook, there's one for your every need! Don't worry if you couldn't quite take it all in: we will go through all of these in detail in the next few modules.
```
> Don't worry if you couldn't quite take it all in: we will go through all of these in detail in the next few modules.

See lesson decoupling note above, we generally try to avoid this

thomastuts commented:
Thanks for the update @kadikraman, this is awesome! Some general feedback:
- Even though an overwhelming majority of our users will probably use (or have used) Facebook, I think using Facebook's UI as an example here might be a little too specific. Can we generalise the examples a bit more by omitting the specific Facebook mentions? We should definitely keep the `ChatBox` example though, it's great for explaining the lifecycle methods. 🎉 
- I think we can add some kind of General 'lifecycle hooks' section right before the `Mounting` part. We can then move the paragraph about rendering in there, as well as serve as a prelude to the lifecycle hooks (we can explain more in-depth that we can 'hook' into these, and React will call the methods when appropriate).
- Should we move the `Unmounting` part after the `Updating` part? I think that would make more sense to the students (more of a natural progression). The rest of the lesson already uses this 'evolution' from creation -> updating -> deletion too.
- When writing content, can you add line breaks after ±100-120 characters to break up paragraphs into separate lines? This makes the reviewing process a lot easier (since I can now only comment on an entire paragraph as it's wrapped to one line), as well as make the content readable in all editors (for example, I don't use text wrapping, so I'd have to scroll for ages in my IDE, but I digress)

Thanks! 💣

thomastuts commented:
@kadikraman 

I hope I'm not too late with this, but the more I think about it, the less my 'remove the Facebook-specific stuff' makes sense - we use other platforms to sometimes describe functionality too, so there's no need to specifically omit Facebook. I _would_ reword the bit about 'Facebook created React..' as I mentioned above in the diff comments, but other than that, the chat box examples are fine.

If you did change things up to be less Facebook-specific, that's fine too, obviously! Either way will work great.

Commit:
Update based on review comments

- remove references to "other lessons"
- remove confusing example updating state based on props
- add a section for lifecycle hooks and rendering

kadikraman commented:
@thomastuts 

I did remove the reference to Facebook, but the rest of the analogy is still there.

I just read your comment about possibly using a life/death analogy... I did actually start out with that, but it seemed pretty grim so I dropped it. Do you think it would still be good to add it? I was trying to think of a life and death of something that we don't care about that much... like a tree or a bug? 🐛 

Wow, what kind of editor are you using that doesn't wrap text? I've never thought of that! Though I guess when writing JavaScript, if you're going over 150 characters, you have bigger problems to deal with anyway 🙈

I've added a lot more paragraphs, but I don't quite want to put every sentence in a  separate line. Is this okay with you? If not, could you link me to an example of a lesson that's paragraphed in a way you like and I can mimic them.

thomastuts commented:
Yeah what you're suggesting using something that people won't have very strong feelings for is great if we do decide to go with that metaphor (e.g. a tree would be great: `componentWillUnmount` would be used to make the birds on its branches fly off before the tree falls over or something?).

I use Webstorm, but I explicitly turn off wrapping to avoid having to deal with super long lines — if you take a look at your GitHub emails (if you have those turned on), for example, you'll see the code on one really long line too. Terminal output is another biggie, it doesn't play well with long lines either in my experience. As a general rule, my right margin is set 120 characters. Obviously it's not a catastrophe if you hit 121 characters but I try to limit things where possible (here's an [example](https://raw.githubusercontent.com/learn-co-curriculum/react-updating-state-lab/84100a77b415a92ace4ebcf523aae637fcdc9a82/README.md)).

Commit:
Add line breaks after every 120 characters

Commit:
Added an Oak tree analogy

thomastuts committed:
Fix typos and improve readability

thomastuts committed:
Use backticks where appropriate

thomastuts commented:
@kadikraman Merging, thanks for the awesome work!
